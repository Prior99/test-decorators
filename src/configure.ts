export type BeforeEach = () => Promise<void> | void

export type BeforeAll = (fn: BeforeAllFunction) => Promise<void> | void
export type BeforeAllFunction = (done?: () => void) => Promise<void> | void

export type It = (name: string, fn: ItFunction) => void
export type ItFunction = (done?: () => void) => Promise<void> | void

export type Describe = (name: string, fn: DescribeFunction) => void
export type DescribeFunction = () => Promise<void> | void

/**
 * Options provided to `configure` for setting up the test framework.
 */
export interface ConfigureOptions {
    /**
     * The function called once per suite before all tests.
     */
    beforeAll?: BeforeAll,
    /**
     * The function called once per suite before each test.
     */
    beforeEach?: BeforeEach,
    /**
     * The function called for each test.
     */
    it: It,
    /**
     * The function called for each test if `only` is set to `true`.
     */
    itOnly?: It,
    /**
     * The function called for each suite.
     */
    describe: Describe,
    /**
     * The function called for each suite if `only` is set to `true`.
     */
    describeOnly?: Describe
}

export let configuration: ConfigureOptions

/**
 * Call this function in the setup code of your test runner to provide the functions
 * to be called for each suite and each test.
 */
export function configure(options: ConfigureOptions) {
    Object.assign(configuration, options)
}

declare var global: any

/**
 * Initialize the configuration with some default values working out of the box for jest an mocha.
 */
export function initializeConfiguration() {
    // Shamelessly adapted from https://github.com/purposeindustries/window-or-global
    // istanbul ignore next
    const globalOrWindow = (typeof self === "object" && self.self === self && self) ||
        (typeof global === "object" && global.global === global && global)

    configuration = {
        beforeEach: (globalOrWindow as any).beforeEach,
        beforeAll: (globalOrWindow as any).beforeAll,
        it: (globalOrWindow as any).it,
        describe: (globalOrWindow as any).describe,
        describeOnly: (globalOrWindow as any).describe.only,
        itOnly: (globalOrWindow as any).it.only,
    }
}

initializeConfiguration()
