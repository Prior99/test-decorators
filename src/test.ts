import "reflect-metadata"
import { TestOptionsInput, parseTestOptions, isTestOptionsInput } from "./options"
import { configuration } from "./configure"

export type TestFunction = (instance: any) => void

export function getTests(target: any): TestFunction[] {
    const tests = Reflect.getMetadata(`test-decorator:tests`, target)
    if (typeof tests !== "undefined") { return tests }
    const newTests: TestFunction[] = []
    Reflect.defineMetadata(`test-decorator:tests`, newTests, target)
    return newTests
}

export type TestDecorator =
    (target: Object, property: string | symbol, descriptor: PropertyDescriptor) => PropertyDescriptor

export function test<Params>(
    target: Object,
    property: string | symbol,
    descriptor: PropertyDescriptor,
): PropertyDescriptor
export function test<Params>(options?: TestOptionsInput<Params>): TestDecorator
/**
 * Decorate a method with `@test` or `@test("name of the test")` or @test({ ... })` to
 * automatically have it run as a test inside an `it` call.
 *
 * @param options Can be provided or left out, so the decorator can be called with `@test()` or
 *                like `@test`. The name of the test or a configuration object. @see TestOptions
 *
 * @return The decorated method.
 */
export function test<Params>(
    arg1: Object | TestOptionsInput<Params>,
    arg2?: string | symbol,
    arg3?: PropertyDescriptor,
) {
    // Load the `it` and `itOnly` functions from the configuration which is stored singleton.
    // It might have been configured to overwrite the mocha ones.
    const { it, itOnly } = configuration
    if (typeof it !== "function") {
        throw new Error("'it' not found. Did you call 'configure'?")
    }
    const options = isTestOptionsInput(arg1) ? parseTestOptions(arg1) : {
        name: () => (arg2 as string),
        params: undefined,
        only: false,
    }
    const { name, params, only } = options
    const decorator: TestDecorator = (target, property, descriptor) => {
        const tests = getTests(target.constructor)
        const itFunction = only ? itOnly : it
        if (only && typeof itOnly !== "function") {
            throw new Error("Call to 'itOnly' occured but it was not configured.")
        }
        // If `params` was not specified, call `it` once with no parameters.
        if (!params) {
            tests.push(instance => {
                // prevents super-class from executing the tests of extending classes
                if (!(instance instanceof target.constructor)) return;
                itFunction(name(), async (...args: any[]) => await descriptor.value.apply(instance, args))
            })
            return descriptor
        }
        // Call `it` once for all parameters configured in `params`.
        tests.push(...params.map(param => {
            return (instance: any) => {
                // prevents super-class from executing the tests of a child class
                // while also preventing the child class from executing parent tests
                if (!(instance.constructor.name === target.constructor.name)) return;
                itFunction(
                    name(param),
                    async (...args: any[]) => await descriptor.value.apply(instance, [param, ...args]),
                )
            }
        }))
        return descriptor
    }
    // Function was used as a decorator with a configuration (`@test(...)`).
    if (isTestOptionsInput(arg1)) {
        return decorator
    }
    // Function was used as a decorator without a configuration (`@test`).
    return decorator(arg1, arg2, arg3)
}
