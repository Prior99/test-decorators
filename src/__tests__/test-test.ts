import { getTests, test as testDecorator } from "../test"
import { configure, It } from "../configure"

describe("`test`", () => {
    let mockIt: jest.Mock<It>
    let mockItOnly: jest.Mock<It>

    beforeEach(() => {
        mockIt = jest.fn()
        mockItOnly = jest.fn()
        configure({ it: mockIt, itOnly: mockItOnly, describe })
    })

    test("without a configuration", () => {
        class A {
            @testDecorator
            private testSomething() { return }
        }
        expect(getTests(A).length).toBe(1)
        getTests(A)[0](undefined)
        expect(mockIt.mock.calls).toMatchSnapshot()
    })

    test("with a name", () => {
        class A {
            @testDecorator("some name")
            private testSomething() { return }
        }
        expect(getTests(A)).toMatchSnapshot()
    })

    test("with a configuration", () => {
        class A {
            @testDecorator({ name: "some name" })
            private testSomething() { return }
        }
        expect(getTests(A).length).toBe(1)
        getTests(A)[0](undefined)
        expect(mockIt.mock.calls).toMatchSnapshot()
    })

    test("with no `it` function provided", () => {
        configure({
            it: undefined,
            describe,
        })
        expect(() => {
            class A {
                @testDecorator
                private testSomething() { return }
            }
        }).toThrowErrorMatchingSnapshot()
    })

    test("with `only` set to `true` but no `itOnly` function provided", () => {
        configure({
            it,
            describe,
            itOnly: undefined,
        })
        expect(() => {
            class A {
                @testDecorator({ name: "some name", only: true })
                private testSomething() { return }
            }
        }).toThrowErrorMatchingSnapshot()
    })

    test("with `params`", () => {
        class A {
            @testDecorator({
                name: (param) => `some name #${param}`,
                params: [1, 2, 3],
            })
            private testSomething() { return }
        }
        const tests = getTests(A)
        expect(tests.length).toBe(3)
        tests.forEach(storedTest => storedTest(undefined))
        expect(mockIt.mock.calls).toMatchSnapshot()
    })

    test("calling the test when configured with `params`", () => {
        const mockTestImpl = jest.fn()
        class A {
            @testDecorator({
                name: (param) => `some name #${param}`,
                params: [
                    { a: 1, b: 2 },
                    { a: 3, b: 4 },
                ],
            })
            private testSomething(...args: any[]) { mockTestImpl(...args) }
        }
        const tests = getTests(A)
        tests.forEach(storedTest => storedTest(undefined))
        mockIt.mock.calls.forEach(call => call[1]())
        expect(mockTestImpl.mock.calls).toMatchSnapshot()
    })

    test("calling the test when configured without `params`", () => {
        const mockTestImpl = jest.fn()
        class A {
            @testDecorator
            private testSomething(...args: any[]) { mockTestImpl(...args) }
        }
        const tests = getTests(A)
        tests[0](undefined)
        mockIt.mock.calls[0][1]()
        expect(mockTestImpl.mock.calls).toMatchSnapshot()
    })
})
