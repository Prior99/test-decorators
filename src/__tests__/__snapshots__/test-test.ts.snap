// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`test\` calling the test when configured with \`params\` 1`] = `
Array [
  Array [
    Object {
      "a": 1,
      "b": 2,
    },
  ],
  Array [
    Object {
      "a": 3,
      "b": 4,
    },
  ],
]
`;

exports[`\`test\` calling the test when configured without \`params\` 1`] = `
Array [
  Array [],
]
`;

exports[`\`test\` with \`only\` set to \`true\` but no \`itOnly\` function provided 1`] = `"Call to 'itOnly' occured but it was not configured."`;

exports[`\`test\` with \`params\` 1`] = `
Array [
  Array [
    "some name #1",
    [Function],
  ],
  Array [
    "some name #2",
    [Function],
  ],
  Array [
    "some name #3",
    [Function],
  ],
]
`;

exports[`\`test\` with a configuration 1`] = `Array []`;

exports[`\`test\` with a name 1`] = `
Array [
  [Function],
]
`;

exports[`\`test\` with no \`it\` function provided 1`] = `"'it' not found. Did you call 'configure'?"`;

exports[`\`test\` without a configuration 1`] = `Array []`;
